bee-ledger/src/workers/storage.rs

```diff
@@ -0,0 +1,496 @@
+        .map_err(|e| Error::Storage(Box::new(e)))
+}
+
+pub(crate) async fn is_output_unspent<B: StorageBackend>(storage: &B, output_id: &OutputId) -> Result<bool, Error> {
+    Exist::<Unspent, ()>::exist(storage, &(*output_id).into())
+        .await
+        .map_err(|e| Error::Storage(Box::new(e)))
+}
+
+pub(crate) async fn insert_treasury_output<B: StorageBackend>(
+    storage: &B,
+    treasury_output: &TreasuryOutput,
+) -> Result<(), Error> {
+    Insert::<(bool, TreasuryOutput), ()>::insert(storage, &(false, treasury_output.clone()), &())
+        .await
+        .map_err(|e| Error::Storage(Box::new(e)))
+}
+
+pub(crate) fn insert_treasury_output_batch<B: StorageBackend>(
+    storage: &B,
+    batch: &mut <B as BatchBuilder>::Batch,
+    treasury_output: &TreasuryOutput,
+) -> Result<(), Error> {
+    Batch::<(bool, TreasuryOutput), ()>::batch_insert(storage, batch, &(false, treasury_output.clone()), &())
+        .map_err(|e| Error::Storage(Box::new(e)))
+}
+
+pub(crate) fn delete_treasury_output_batch<B: StorageBackend>(
+    storage: &B,
+    batch: &mut <B as BatchBuilder>::Batch,
+    treasury_output: &TreasuryOutput,
+) -> Result<(), Error> {
+    Batch::<(bool, TreasuryOutput), ()>::batch_delete(storage, batch, &(false, treasury_output.clone()))
+        .map_err(|e| Error::Storage(Box::new(e)))
+}
+
+pub(crate) fn spend_treasury_output_batch<B: StorageBackend>(
+    storage: &B,
+    batch: &mut <B as BatchBuilder>::Batch,
+    treasury_output: &TreasuryOutput,
+) -> Result<(), Error> {
+    Batch::<(bool, TreasuryOutput), ()>::batch_insert(storage, batch, &(true, treasury_output.clone()), &())
+        .map_err(|e| Error::Storage(Box::new(e)))?;
+    Batch::<(bool, TreasuryOutput), ()>::batch_delete(storage, batch, &(false, treasury_output.clone()))
+        .map_err(|e| Error::Storage(Box::new(e)))
+}
+
+pub(crate) fn unspend_treasury_output_batch<B: StorageBackend>(
+    storage: &B,
+    batch: &mut <B as BatchBuilder>::Batch,
+    treasury_output: &TreasuryOutput,
+) -> Result<(), Error> {
+    Batch::<(bool, TreasuryOutput), ()>::batch_insert(storage, batch, &(false, treasury_output.clone()), &())
+        .map_err(|e| Error::Storage(Box::new(e)))?;
+    Batch::<(bool, TreasuryOutput), ()>::batch_delete(storage, batch, &(true, treasury_output.clone()))
+        .map_err(|e| Error::Storage(Box::new(e)))
+}
+
+/// Fetches the unspent treasury output from the storage.
+pub async fn fetch_unspent_treasury_output<B: StorageBackend>(storage: &B) -> Result<TreasuryOutput, Error> {
+    match Fetch::<bool, Vec<TreasuryOutput>>::fetch(storage, &false)
+        .await
+        .map_err(|e| Error::Storage(Box::new(e)))?
+    {
+        Some(outputs) => {
+            match outputs.len() {
+                // There has to be an unspent treasury output at all time.
+                0 => panic!("No unspent treasury output found"),
+                // Indexing is fine since length is known
+                1 => Ok(outputs[0].clone()),
```

Comment:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove allows you to remove the entry you want without a `.clone()`.